File: relay_server.py
import os
import sys
import asyncio
import urllib.parse
from http import HTTPStatus
import websockets

import shutil

PORT = int(os.environ.get("PORT", 8080))

def get_ffmpeg_path():
    if os.environ.get("FFMPEG_PATH"):
        return os.environ.get("FFMPEG_PATH")
    if shutil.which("ffmpeg"):
        return "ffmpeg"
    fallback = r"C:\Program Files\Streamlabs OBS\resources\node_modules\ffmpeg-ffprobe-static\ffmpeg.exe"
    if os.path.exists(fallback):
        return fallback
    return "ffmpeg"

async def handle_client(websocket):
    parsed = urllib.parse.urlparse(websocket.request.path)
    if parsed.path != "/live":
        await websocket.close(1008, "Invalid endpoint. Use /live")
        return

    query = urllib.parse.parse_qs(parsed.query)
    stream_keys = query.get("key", [])
    if not stream_keys or not stream_keys[0]:
        print("[Relay] Connection rejected: Missing YouTube stream key.", flush=True)
        await websocket.close(1008, "Stream key required")
        return

    stream_key = stream_keys[0]
    rtmp_url = f"rtmp://a.rtmp.youtube.com/live2/{stream_key}"
    print(f"[Relay] New client connected. Forwarding stream to YouTube RTMP...", flush=True)

    ffmpeg_bin = get_ffmpeg_path()
    ffmpeg_args = [
        ffmpeg_bin,
        "-i", "-",               # Read incoming WebM chunks from stdin
        "-c:v", "copy",          # Passthrough video
        "-c:a", "aac",           # Transcode audio to AAC
        "-b:a", "128k",
        "-ar", "44100",
        "-f", "flv",             # Container format for YouTube RTMP
        rtmp_url
    ]

    try:
        process = await asyncio.create_subprocess_exec(
            *ffmpeg_args,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL
        )
    except Exception as e:
        print(f"[Relay] Failed to start FFmpeg: {e}", flush=True)
        await websocket.close(1011, "FFmpeg execution error")
        return

    try:
        async for message in websocket:
            if isinstance(message, bytes) and process.stdin:
                try:
                    process.stdin.write(message)
                    await process.stdin.drain()
                except Exception as write_err:
                    print(f"[Relay] Error writing to FFmpeg: {write_err}", flush=True)
                    break
    except websockets.exceptions.ConnectionClosed:
        pass
    finally:
        print("[Relay] Client disconnected. Closing FFmpeg stream...", flush=True)
        if process.stdin:
            try:
                process.stdin.close()
                await process.stdin.wait_closed()
            except Exception:
                pass
        try:
            process.terminate()
            await process.wait()
        except Exception:
            pass

async def process_request(connection, request):
    parsed = urllib.parse.urlparse(request.path)
    if parsed.path != "/live":
        # Return plain HTTP response for root or status checks
        return connection.respond(
            HTTPStatus.OK,
            "Meet Recorder RTMP Relay Server is running.\n"
        )
    return None

async def main():
    print(f"[Relay Server] Listening on port {PORT}", flush=True)
    print(f"[Relay Server] WebSocket endpoint: ws://localhost:{PORT}/live", flush=True)
    async with websockets.serve(
        handle_client,
        "0.0.0.0",
        PORT,
        process_request=process_request
    ):
        await asyncio.Future()  # run forever

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n[Relay Server] Stopped.", flush=True)

File: relay-server.js
const { createServer } = require('http');
const { parse } = require('url');
const { WebSocketServer } = require('ws');
const { spawn } = require('child_process');

const PORT = process.env.PORT || 8080;
const server = createServer((req, res) => {
  res.writeHead(200, { 'Content-Type': 'text/plain' });
  res.end('Meet Recorder RTMP Relay Server is running.\n');
});

const wss = new WebSocketServer({ noServer: true });

server.on('upgrade', (request, socket, head) => {
  const { pathname } = parse(request.url);

  if (pathname === '/live') {
    wss.handleUpgrade(request, socket, head, (ws) => {
      wss.emit('connection', ws, request);
    });
  } else {
    socket.destroy();
  }
});

wss.on('connection', (ws, req) => {
  const { query } = parse(req.url, true);
  const streamKey = query.key;

  if (!streamKey) {
    console.warn('[Relay] Connection rejected: Missing YouTube stream key.');
    ws.close(1008, 'Stream key required');
    return;
  }

  const rtmpUrl = `rtmp://a.rtmp.youtube.com/live2/${streamKey}`;
  console.log(`[Relay] New client connected. Forwarding stream to YouTube RTMP...`);

  // Spawn ffmpeg to transcode/repackage WebM (Opus) -> FLV (AAC) for RTMP
  const ffmpegArgs = [
    '-i', '-',                          // Read incoming WebM chunks from stdin
    '-c:v', 'copy',                     // Direct passthrough for video (low CPU)
    '-c:a', 'aac',                      // RTMP requires AAC audio
    '-b:a', '128k',                     // 128k AAC audio bitrate
    '-ar', '44100',                     // Standard RTMP sample rate
    '-f', 'flv',                        // YouTube RTMP container format
    rtmpUrl
  ];

  const { existsSync } = require('fs');
  const getFFmpegPath = () => {
    if (process.env.FFMPEG_PATH) return process.env.FFMPEG_PATH;
    const fallback = 'C:\\Program Files\\Streamlabs OBS\\resources\\node_modules\\ffmpeg-ffprobe-static\\ffmpeg.exe';
    if (existsSync(fallback)) return fallback;
    return 'ffmpeg';
  };

  const ffmpegProcess = spawn(getFFmpegPath(), ffmpegArgs);

  ffmpegProcess.stderr.on('data', (data) => {
    // Uncomment for debugging FFmpeg stream status:
    // console.log(`[FFmpeg] ${data.toString()}`);
  });

  ffmpegProcess.on('error', (err) => {
    console.error('[FFmpeg Process Error]', err);
    ws.close(1011, 'FFmpeg execution error');
  });

  ffmpegProcess.on('close', (code, signal) => {
    console.log(`[FFmpeg] Closed with code ${code}, signal ${signal}`);
  });

  // Pipe binary ArrayBuffers directly into FFmpeg stdin
  ws.on('message', (chunk, isBinary) => {
    if (isBinary && ffmpegProcess.stdin.writable) {
      ffmpegProcess.stdin.write(chunk);
    }
  });

  ws.on('close', () => {
    console.log('[Relay] Client disconnected. Closing FFmpeg stream...');
    if (ffmpegProcess.stdin.writable) {
      ffmpegProcess.stdin.end();
    }
    ffmpegProcess.kill('SIGINT');
  });

  ws.on('error', (err) => {
    console.error('[WebSocket Client Error]', err);
    if (ffmpegProcess.stdin.writable) {
      ffmpegProcess.stdin.end();
    }
    ffmpegProcess.kill('SIGINT');
  });
});

server.listen(PORT, () => {
  console.log(`[Relay Server] Listening on port ${PORT}`);
  console.log(`[Relay Server] WebSocket endpoint: ws://localhost:${PORT}/live`);
});